fixes #26186 - add react-router foundation

glekner opened the pull request:
Splitting the AuditsPage API PR #6469  into smaller ones
This PR adds foundation for react-router routing and navigation
<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [Translating section in the guide]
(https://projects.theforeman.org/projects/foreman/wiki/Translating)
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* Be nice and respectful

We are running bots that will poke you if you miss an item from the list :-)

--->

glekner committed:
fixes #26186 - add react-router foundation

glekner commented on webpack/assets/javascripts/react_app/components/Layout/__tests__/__snapshots__/LayoutSelectors.test.js.snap:
```diff
@@ -98,17 +98,19 @@ Array [
         "href": "/",
         "isDivider": false,
-        "onClick": null,
+        "onClick": undefined,
```
@ohadlevy the changes here are only from `null` to `undefined` and adding url for each child, seems ok.

ohadlevy commented on webpack/assets/javascripts/react_app/routes/index.js:
```diff
@@ -0,0 +1,35 @@
+    <Route
+      render={({ location }) => {
+        if (currentLocation && currentLocation.pathname !== location.pathname) {
+          window.Turbolinks.visit(location.pathname);
```
from some reason, I"m seeing two turbolink events, when the first one is cancled followed by a real one, it seems its somehow registered multiple times
![image](https://user-images.githubusercontent.com/20167/53662113-c7ee6480-3c6a-11e9-974b-ae1e4db26a65.png)

ohadlevy reviewed (dismissed):
this PR seems to work (hard to test it without react routes).
I've left a question regarding multiple on click events firing when navigating between pages.

I believe we already have the missing packages (as they are used by katello?) - @ekohl would appreciate your ack.

ekohl commented:
> I believe we already have the missing packages (as they are used by katello?) - @ekohl would appreciate your ack.

We'll need to move over from one repo to another. The history package also needs be packaged. I'll take a look at this.

glekner commented:
removed the fallback route @ohadlevy

tbrisker commented on app/views/layouts/base.html.erb:
```diff
@@ -49,11 +49,13 @@
       <% end %>
-      <div class="container-fluid container-pf-nav-pf-vertical nav-pf-persistent-secondary">
-      <%# mounting layout after the container to change its width if the navbar is collapsed %>
-      <%= mount_layout %> 
```
Should this helper be removed than?

glekner replied:
removed

tbrisker commented on app/views/layouts/base.html.erb:
```diff
@@ -49,11 +49,13 @@
+        <%= mount_react_app %>
       <% end %>
-      <div class="container-fluid container-pf-nav-pf-vertical nav-pf-persistent-secondary">
-      <%# mounting layout after the container to change its width if the navbar is collapsed %>
```
is this comment no longer relevant?

glekner replied:
we handle the collapsed state in the app now, so yes

tbrisker commented on app/views/layouts/base.html.erb:
```diff
@@ -49,11 +49,13 @@
       <% cache(TopbarSweeper.fragment_name, :skip_digest => true) do %>
-        <div id="layout"></div>
+        <div id="react-app-root"></div>
+        <%= mount_react_app %>
```
this should now cache the `react_app_data` i think, is this on purpose? (i think before we actually lost the caching when the data moved outside of the cache block - originally the navbars were cached to save recalculating the user's permissions on every page)

glekner replied:
I wasn't sure so i kept it. can it be deleted?

tbrisker replied:
I think you misunderstood my comment - we should cache this data, otherwise we need to recalculate permissions etc every time a page reloads (which would be less with this but still). Another option is to move this caching to the serverside. keep in mind that we might need to change the TopbarSweeper implementation to properly handle the new layout.

glekner replied:
changing the TopbarSweeper implementation is out of scope, and out of my knowledge right now. Maybe we should create an issue for this.
afaik The Menu items are [cached in the client](https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/Layout/LayoutSelectors.js#L14) but i guess you are referring to something else with the permissions. either way I think we should handle this in another PR

tbrisker commented on webpack/assets/javascripts/react_app/components/Layout/LayoutSelectors.js:
```diff
@@ -26,16 +28,20 @@ const patternflyItems = (data, currentLocation, currentOrganization) => {
     const childrenArray = [];
     item.children.forEach(child => {
+      const isReact =
+        pages.filter(route => route.path === child.url).length > 0;
```
You can use [Array.prototype.some()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some) here, instead of filter which creates a new array with all matching elements.

tbrisker commented on webpack/assets/javascripts/react_app/components/Layout/LayoutSelectors.js:
```diff
@@ -26,16 +28,20 @@ const patternflyItems = (data, currentLocation, currentOrganization) => {
-        preventHref: false,
-        onClick: child.onClick ? () => child.onClick() : null,
+        href: child.url || '#',
+        preventHref: !!isReact,
```
no need for `!!`, `isReact` should be boolean already

tbrisker commented on webpack/assets/javascripts/react_app/components/Layout/LayoutSelectors.js:
```diff
@@ -26,16 +28,20 @@ const patternflyItems = (data, currentLocation, currentOrganization) => {
+        pages.filter(route => route.path === child.url).length > 0;
+
       const childObject = {
+        url: child.url,
```
why do we need both `url` and `href` attributes? seems like duplication

tbrisker commented on webpack/assets/javascripts/react_app/components/Layout/LayoutSelectors.js:
```diff
@@ -26,16 +28,20 @@ const patternflyItems = (data, currentLocation, currentOrganization) => {
-        onClick: child.onClick ? () => child.onClick() : null,
+        href: child.url || '#',
+        preventHref: !!isReact,
+        onClick: isReact ? noop : child.onClick,
```
why is this change needed? isn't the onClick used when non-default behaviour is needed for the menu element? (do we even use it anywhere? I couldn't find any in a quick search)

glekner replied:
its for future use cases where plugins or other consumers want another functionality besides the ref. the `noop` is for the PF navbar, it needs to receive a function in its condition in order to preventHref.

glekner replied:
@tbrisker i guess this be replaced with
`onClick: child.onClick || noop`

tbrisker commented on webpack/assets/javascripts/react_app/routes/index.js:
```diff
@@ -0,0 +1,24 @@
+            'rails-application-content'
+          );
+          if (railsApplicationContent) {
+            railsApplicationContent.remove();
```
Is there a way of perhaps having both implementation using the same div instead of duplicating it and then removing?

sharvit replied:
It's not trivial, you can't render rails content into react component and you can only control the rails content by using the `element`.

imo this function should be in a different file with unit-testing.

tbrisker commented on webpack/assets/javascripts/react_app/history.js:
```diff
@@ -0,0 +1,3 @@
+import { createBrowserHistory } from 'history';
+
+export default createBrowserHistory();
```
Why add this file instead of directly importing it in `ReactApp.js`?

sharvit replied:
* We might want to import it from different files.
* We might want to add some extra configuration here.

tbrisker reviewed (dismissed):
I did an initial pass over the code (now with just 143 non-whitespace loc change it's actually possible...).
See comments inline. 
Also, what is `react_page.html` for? It seems to be an empty file that was mistakenly added to the commit.

sharvit commented on webpack/assets/javascripts/react_app/routes/index.js:
```diff
@@ -0,0 +1,24 @@
+
+export const pages = []; // React Pages
+
+export default (
```
You are returning a static jsx instead a component.

imo it should be a component called `ApplicationRoutesSwitcher` and `pages` should be called `route` and live in a separate file.

sharvit commented on webpack/assets/javascripts/react_app/routes/index.js:
```diff
@@ -0,0 +1,24 @@
+export default (
+  <Switch>
+    {pages.map(({ render, ...props }) => (
+      <Route
```
I think that if you will add a turbolinks fallback route, you can render react-router-links in the layout so it will only use the react-router. Here, if you can't resolve the route with react, you can use turbolinks.

It will allow you to remove some code that handles `href` and `onClick` and pushing new state to the `history` object.

glekner replied:
I prefer to do this in a future PR, unless it solves the mount/unmount of the layout?

glekner commented:
Added `onCollapse`, `onExpand` for the layout.
Turns out the PF-VerticalNav finds the first container of the app and inject the `collapsed-nav` className if the navbar gets collapsed. We now have 2 containers, one inside the layout and one outside for the rails content. The one outside the layout doesn't change dynamically anymore so we have to manage it in our app, which is also recommended by the VerticalNavbar author :)

<img width="768" alt="screenshot 2019-03-04 14 16 38" src="https://user-images.githubusercontent.com/24938324/53732959-23496e00-3e88-11e9-8101-3b1d8fae7e61.png">

tbrisker commented on webpack/assets/javascripts/react_app/components/Layout/LayoutSelectors.js:
```diff
@@ -26,16 +28,18 @@ const patternflyItems = (data, currentLocation, currentOrganization) => {
   data.forEach(item => {
     const childrenArray = [];
     item.children.forEach(child => {
+      const isReact = routes.some(route => route.path === child.url).length > 0;
```
you don't need the `.length >0` part, `.some` already returns a boolean

timogoebel commented:
I'm wondering how this is supposed to work? I'd expect that there would be a dedicated controller (`ReactController`) that would render the main react div. I'm surprised that the main react div is actually defined on the layout, e.g. rendered on every page.

glekner commented:
@timogoebel not sure I understand. the layout doesn't change, the `Switch` is inside it. We have two containers for the migration process

@tbrisker @sharvit can you ack?

ekohl commented:
Katello already has react-router present. Will this break those pages? If so, is there a corresponding PR?

Something I also noticed while debugging katello: if you get a route wrong, it renders an empty page. This was horrible to debug. Is there something that we could add here to make that process easier? Especially since there are no automated tests to indicate something broke. The result was that in Katello the subscriptions page was broken for 4 releases and I greatly worry that Foreman will have a similar degradation in quality.

timogoebel commented:
> @timogoebel not sure I understand. the layout doesn't change, the `Switch` is inside it. We have two containers for the migration process

Let me phrase a question: How does this work? How does it integrate with the existing rails UI?

glekner commented:
@timogoebel 
All menu items in the navigation have an `href` which Turbolinks listens to and triggers a visit when clicked, rendering content from the Rails view.
With the new changes, if a View is in the routes [array](https://github.com/theforeman/foreman/pull/6522/files#diff-fc6f587e7a5f5e069ab25cbd003fa30dR41) it will [prevent](https://github.com/theforeman/foreman/pull/6522/files#diff-fc6f587e7a5f5e069ab25cbd003fa30dR41) the Turbolinks default behavior. Instead, The Switch will render it, assuming it has a route for it.
Before rendering, the Switch checks if the Rails container is mounted, if so it will remove it.

Let me know if you need more info

glekner commented:
@ekohl we discussed the behaviour of both routers in katello and puglins in general with foreman, we thought about making a route registry along with the component registry, maybe @sharvit can elaborate on this.

Right now going to an unknown route (e.g. /noroute) i'm getting a Rails GET error, which seems to be the wanted behaviour?
If you mean going to a view that has a Switch of its own (e.g. /audits/noroute) then the audits Switch should implement a fallback route.

ekohl commented:
> Right now going to an unknown route (e.g. /noroute) i'm getting a Rails GET error, which seems to be the wanted behaviour?

That sounds like the correct behaviour to me. Clear errors make life easier.

>  If you mean going to a view that has a Switch of its own (e.g. /audits/noroute) then the audits Switch should implement a fallback route.

If they use a correct regex (`/audits/[0-9]+`) then it should already fall back to Rails GET error when passing in `/audits/unknown`, correct? Of course it should still handle 404 responses when an unknown ID is passed in.

timogoebel commented:
> Let me know if you need more info

@glekner: Thanks. That helps a lot. I still don't get how this works if I access a route for a page that is react only. Which rails controller will render this? But maybe this is not even implemented here.

waldenraines commented:
> Katello already has react-router present. Will this break those pages? If so, is there a corresponding PR?

It breaks the old sync management page:

```
undefined local variable or method `mount_react_app' for #<#<Class:0x00007fe418810db8>:0x00007fe41880a710>
Did you mean?  mount_react_component
```

So changes will be required to fix that.

waldenraines commented:
> It breaks the old sync management page:

See https://github.com/Katello/katello/pull/8002 for the fix.

glekner commented:
@sharvit @tbrisker ack please?

timogoebel commented:
@glekner: My questions are still unanswered.

glekner commented:
@timogoebel sorry i missed your question.

I hope i fully understood this. for example we want to migrate the audits page to react, so the audits controller inherits from the new react controller which has a new layout `react_application.html.erb` which **doesn't** render

` <%= render :partial => 'layouts/application_content' %>`

so all the react pages will just stop rendering serverside content and let the router do the job

glekner commented:
> If they use a correct regex (`/audits/[0-9]+`) then it should already fall back to Rails GET error when passing in `/audits/unknown`, correct? Of course it should still handle 404 responses when an unknown ID is passed in.

@ekohl correct, it falls back to Rails GET error

timogoebel commented:
@glekner: Thanks, that answers my questions!

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/Layout/LayoutHelper.js:
```diff
@@ -119,3 +119,20 @@ const createLocationItem = locations => {
 };
+
+export const onCollapse = () => {
+  document.getElementById('rails-app-content').classList.add('collapsed-nav');
```
maybe we can update the class through the redux store changes instead + mounting the component with a new className?  same goes for `onExpand`

Ron-Lavi commented on webpack/assets/javascripts/react_app/routes/index.js:
```diff
@@ -0,0 +1,31 @@
+        key={props.path}
+        render={renderProps => {
+          const railsContainer = document.getElementById('rails-app-content');
+          if (railsContainer) railsContainer.remove();
```
just wondering if this is the right "react way"

glekner replied:
quoting @sharvit 

> It's not trivial, you can't render rails content into react component and you can only control the rails content by using the element.

glekner commented:
- rebased and added `LayoutContainer`
- Seperarted legacy and react classnames for containers. Redux now manages the collapsed state for the react container.

@Laviro @sharvit @tbrisker @ohadlevy can we merge?

Ron-Lavi reviewed (dismissed):
Thanks @glekner,
LGTM :+1: can't wait to see the following PRs :)

boaz0 commented:
Good job @glekner and only 31 files! Great job.
LGTM :+1:

mmoll commented:
[test katello]

mmoll commented:
@glekner please rebase

ohadlevy commented:
@glekner whats the status here? @sharvit @amirfefer can you ack / review? thanks!

glekner commented:
@ohadlevy we agreed to merge this after branching since its foundation, the PR is ready though.

ohadlevy commented:
@glekner it seems that there are a few minor issues that @tbrisker raised that are not addressed?

glekner commented:
I think I addressed all of @tbrisker's suggestions, right? @tbrisker

glekner commented:
- Rebased

We agreed to handle server side caching of the layout in a future PR,
so I think we're good to merge. can you ack @tbrisker ?

Ron-Lavi commented:
@glekner can you fix `prettier` issues ?

glekner commented:
well this is odd, lint is passing locally. can you take a look @sharvit ?
https://travis-ci.org/theforeman/foreman/jobs/519867963

glekner commented:
fixed travis

glekner commented:
@ohadlevy @tbrisker 
After discussing with @sharvit, I decided to re-add the fallback route, thought it might be ideal that the Router will handle and trigger all the navigations, even those that use Turbolinks. Now `preventHref` is true to all menu items, so the router will trigger a `Turbolinks.visit` when necessary (tested pretty much all cases)

ohadlevy commented:
> > I believe we already have the missing packages (as they are used by katello?) - @ekohl would appreciate your ack.
> 
> We'll need to move over from one repo to another. The history package also needs be packaged. I'll take a look at this.

added history at https://github.com/theforeman/foreman-packaging/pull/3660

ohadlevy commented:
it seems something is breaking katello, from jenkins... 
```
Error Message
ActionView::Template::Error: undefined local variable or method `mount_react_app' for #<#<Class:0x0000000014cf99a0>:0x000000005a5865d8>...
Stacktrace
ActionView::Template::Error: undefined local variable or method `mount_react_app' for #<#<Class:0x0000000014cf99a0>:0x000000005a5865d8>
Did you mean?  mount_react_component
    app/views/layouts/base.html.erb:59:in `block in _2b4e82d15b55346be18074afac0e431e'
    app/views/layouts/base.html.erb:58:in `_2b4e82d15b55346be18074afac0e431e'
    plugin/app/views/katello/layouts/katello.html.erb:23:in `_94a5381f7bb64edfc606a467c668330a'
    app/controllers/concerns/foreman/controller/timezone.rb:10:in `set_timezone'
    app/controllers/concerns/foreman/controller/topbar_sweeper.rb:12:in `set_topbar_sweeper_controller' (ActionView::Template::Error)
/home/jenkins/workspace/test_develop_pr_katello/database/postgresql/ruby/2.5/slave/fast/app/views/layouts/base.html.erb:59    
```

glekner commented:
> it seems something is breaking katello, from jenkins...
> 
> ```
> Error Message
> ActionView::Template::Error: undefined local variable or method `mount_react_app' for #<#<Class:0x0000000014cf99a0>:0x000000005a5865d8>...
> Stacktrace
> ActionView::Template::Error: undefined local variable or method `mount_react_app' for #<#<Class:0x0000000014cf99a0>:0x000000005a5865d8>
> Did you mean?  mount_react_component
>     app/views/layouts/base.html.erb:59:in `block in _2b4e82d15b55346be18074afac0e431e'
>     app/views/layouts/base.html.erb:58:in `_2b4e82d15b55346be18074afac0e431e'
>     plugin/app/views/katello/layouts/katello.html.erb:23:in `_94a5381f7bb64edfc606a467c668330a'
>     app/controllers/concerns/foreman/controller/timezone.rb:10:in `set_timezone'
>     app/controllers/concerns/foreman/controller/topbar_sweeper.rb:12:in `set_topbar_sweeper_controller' (ActionView::Template::Error)
> /home/jenkins/workspace/test_develop_pr_katello/database/postgresql/ruby/2.5/slave/fast/app/views/layouts/base.html.erb:59    
> ```

theres a [katello pr](https://github.com/Katello/katello/pull/8002) waiting to get merged @ohadlevy

ohadlevy commented:
> theres a [katello pr](https://github.com/Katello/katello/pull/8002) waiting to get merged @ohadlevy

I would assume this would no longer needed when we move fully to react router ?

ohadlevy commented:
@glekner please rebase.
@tbrisker anything else for this PR from your POV? thanks.

glekner commented:
rebased

ohadlevy commented on webpack/assets/javascripts/react_app/routes/index.js:
```diff
@@ -0,0 +1,48 @@
+  path: '',
+};
+
+export default AppSwitcher;
```
does it make sense to add it to a storybook?

glekner replied:
don't think so, this is 100% infrastructure

ohadlevy commented on app/helpers/layout_helper.rb:
```diff
@@ -1,8 +1,4 @@
 module LayoutHelper
-  def mount_layout
-    mount_react_component('Layout', "#layout", layout_data.to_json)
```
can we avoid this change so the PR in katello won't be required?

glekner replied:
removed `react_app_helper`

sharvit commented on webpack/assets/javascripts/react_app/components/Layout/Layout.js:
```diff
@@ -22,8 +28,10 @@ class Layout extends React.Component {
+      onCollapse,
     } = this.props;
     if (items.length === 0) fetchMenuItems(data);
+    if (checkCollapsed()) onCollapse();
```
I'm wondering...

Instead of calling `onCollapse` when we want to initiate the state with `isCollapse=true`.
Would it make sense to do it in the initial state creation inside the reducer?

```js
const initialState = Immutable({
  isCollapsed: checkCollapsed(),
});
```

sharvit reviewed (commented):
Awesome work @glekner :+1:

ohadlevy commented on package.json:
```diff
@@ -78,6 +78,7 @@
     "diff": "~3.0.0",
     "file-saver": "^2.0.1",
     "gridster": "^0.5.6",
+    "history": "^4.7.2",
```
do we need it as a top dependency or is it part of what react-router-dom requires?

glekner replied:
top

ohadlevy commented on package.json:
```diff
@@ -104,6 +105,7 @@
     "react-onclickoutside": "^6.6.2",
     "react-password-strength": "^2.4.0",
     "react-redux": "^5.0.6",
+    "react-router-dom": "^4.3.1",
```
this should be added to webpack.vendor.js as its used by Katello as well (same for history if its really required).

glekner commented:
added packages to `webpack.vendor.js` and checking collapsed state in the initial state
